hyperspy/io_plugins/emd.py

```diff
@@ -404,14 +421,581 @@ def log_info(self):
+        self.read_SI_image_stack = read_SI_image_stack
+        with h5py.File(filename, 'r') as f:
+            self.d_grp = f.get('Data')
+            self._check_im_type()
+            self._read_data()
+
+    def _read_data(self):
+        if self.im_type == 'Image':
+            _logger.info('Reading the images')
+            self._read_images()
+        elif self.im_type == 'Spectrum':
+            self._read_spectrums()
+        elif self.im_type == 'SpectrumStream':
+            _logger.info('Reading the spectrum image')
+            self._read_images()
+            self._read_spectrum_image()
+
+    def _check_im_type(self):
+        if 'Image' in self.d_grp:
+            if 'SpectrumImage' in self.d_grp:
+                self.im_type = 'SpectrumStream'
+            else:
+                self.im_type = 'Image'
+        else:
+            self.im_type = 'Spectrum'
+
+    def _read_spectrums(self):
+        self.record_by = 'spectrum'
+        spectrum_grp = self.d_grp.get("Spectrum")
+        self.detector_name = 'EDS'
+        for spectrum_sub_group_key in _get_keys_from_group(spectrum_grp):
+            self.dictionaries.append(
+                self._read_spectrum(spectrum_grp, spectrum_sub_group_key))
+
+    def _read_spectrum(self, spectrum_group, spectrum_sub_group_key):
+        spectrum_sub_group = spectrum_group[spectrum_sub_group_key]
+        dataset = spectrum_sub_group['Data']
+        data = dataset[:, 0]
+
+        original_metadata = _parse_metadata(spectrum_group,
+                                            spectrum_sub_group_key)
+
+        dispersion, offset = self._get_dispersion_offset(original_metadata)
+
+        axes = [{'index_in_array': 0,
+                 'name': 'E',
+                 'offset': offset,
+                 'scale': dispersion,
+                 'size': data.shape[0],
+                 'units': 'keV'}
+                ]
+
+        md = self._get_metadata_dict()
+        md['Signal']['signal_type'] = 'EDS_TEM'
+
+        return {'data': data,
+                'axes': axes,
+                'metadata': md,
+                'original_metadata': original_metadata,
+                'mapping': self._get_mapping()}
+
+    def _read_images(self):
+        self.record_by = 'image'
+        # Get the image data group
+        image_group = self.d_grp.get("Image")
+        # Get all the subgroup of the image data group and read the image for
+        # each of them
+        for image_sub_group_key in _get_keys_from_group(image_group):
+            self.dictionaries.append(
+                self._read_image(image_group, image_sub_group_key))
+
+    def _read_image(self, image_group, image_sub_group_key):
+        """ Return a dictionary ready to parse of return to io module"""
+        image_sub_group = image_group[image_sub_group_key]
+        original_metadata = _parse_metadata(image_group, image_sub_group_key)
+        self.detector_name = original_metadata['BinaryResult']['Detector']
```

Comment:
With the upcoming release of Velox 2.2, we also save Elemental Maps in images. This means that every image here will have SuperX as name.

Reply:
With upcoming release of Velox 2.3, due to technical reasons, the BinaryResult:Detector field is no longer guaranteed to be filled in all cases. If it's not filled in, there will be only 1 detector under "Detectors"; which is the used detector for the acquisition.

Reply:
Thanks for the info. Do you a document describing the format specification?

Reply:
Unfortunately no. The format is still in flux while we're trying to stabilise it more and more. Once we feel it's at a stable point, we will update documentation. Feel free to ask me any questions though; I know all the little details about the EMD file format.

Reply:
Great, thanks!

Reply:
> With the upcoming release of Velox 2.2, we also save Elemental Maps in images. This means that every image here will have SuperX as name.

@robindegen, where is the name of each elemental map saved?

Reply:
@ericpre There's a few ways you can implement this. I'll share the simplest method, because the "real" way requires you to parse almost everything in the file (json text arrays). If you're interested in that approach we should probably do a chat/mail session at some point.

Anyway:
Under the ```/Presentation/Displays/ImageDisplay/``` group there are a whole bunch of text arrays containing Json. Other than the color mix display, there will be a display for each element that was selected in the periodic table, plus for each selected STEM detector. 

The Json looks like this (example):
```
{
    "display": {
        "id": "Co",
        "label": "Co",
        "overlays": "",
        "priority": "27",
        "closable": "false",
        "visible": "true"
    },
    "dataPath": "\/Data\/Image\/9a522c7075e94dcbbd781e55abf9e36d",
    "seriesIndex": "0",
    "zoom": "1",
    "pan": {
        "x": "0",
        "y": "0"
    },
    "angle": "0",
    "cropMode": "Off"
}
```

The ```dataPath``` will point to one of the images, and the ```display.label``` will be the name of the element or STEM detector respectively. 

Note that label is a URL encoded string, which can contain UTF-8 characters. For example it can contain %C2%B5, which translates to µ (Micro).

Again, be aware that this is a little bit of a trick, as this is not really how Velox interprets this data. This should however work for pretty much all images, even for DPC or Ceta experiments. There will always be a somewhat descriptive title under "Label".

Hope this helps.

Reply:
Thanks for your quick reply. I have seen this but I was a bit disappointed to notice that it doesn't tell us what energy line it is.... Is it necessary to parse all the other metadata (it seems that there is table of all elements with the corresponding energy line used for mapping) to just get the energy line?

Reply:
@ericpre You can use the ImageQuantificationRecord dataset for that.  There should be one at ```/Operations/ImageQuantificationOperation/(guid)```. This is again a text array that contains Json. In there you will find ```elementProperties```.

There are element properties (and thus line selections) saved for all elements that Velox supports.
There can only be 1 image per element, and line selections are tied to that element. You can't have a seperate elemental map image for each line for example (Velox doesn't currently support this).

So for a specific element, you can look here at what the line selections are. The selected lines for a specific element are saved as xray transitions (iupac style).

Excerpt from this portion of the Json:
```
  "elementProperties": [
        {
            "element": "1",
            "xrayTransitions": [
                {
                    "L3": "K"
                }
            ],
            "atomicShellFamily": "K",
            "quantify": "true",
            "includeInAutoPeakId": "true",
            "defaultLineSelect": "true"
        },
        {
            "element": "2",
            "xrayTransitions": [
                {
                    "L3": "K"
                }
            ],
            "atomicShellFamily": "K",
            "quantify": "true",
            "includeInAutoPeakId": "true",
            "defaultLineSelect": "true"
        },
        {
            "element": "3",
            "xrayTransitions": [
                {
                    "L3": "K"
                }
            ],
            "atomicShellFamily": "K",
            "quantify": "true",
            "includeInAutoPeakId": "true",
            "defaultLineSelect": "true"
        },
```

Hope this helps.